File: go/0015_3sum.go
package main

import "sort"

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	numslen := len(nums)

	var res [][]int
	for i := 0; i < numslen-2; i++ {
		if i != 0 && nums[i] == nums[i-1] {
			continue
		}
		target := 0 - nums[i]
		for start, end := i+1, numslen-1; start < end; {
			twoSum := nums[start] + nums[end]
			if twoSum == target {
				res = append(res, []int{nums[i], nums[start], nums[end]})
				for start < end && nums[start+1] == nums[start] {
					start++
				}
				for start < end && nums[end-1] == nums[end] {
					end--
				}
				start++
				end--
			} else if twoSum < target {
				for start < end && nums[start+1] == nums[start] {
					start++
				}
				start++
			} else {
				for start < end && nums[end-1] == nums[end] {
					end--
				}
				end--
			}
		}
	}
	return res
}

File: python/0015_3_sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            target = 0 - nums[i]
            left, right = i + 1, len(nums) - 1

            while left < right:
                if target == nums[left] + nums[right]:
                    res.append([nums[i], nums[left], nums[right]])
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1
                elif target < nums[left] + nums[right]:
                    right -= 1
                else:
                    left += 1
        return res
